[18.0][MIG][REF] sale_triple_discount: Consolidate discount in std field

sbejaoui opened the pull request:
This PR includes a refactoring inspired by https://github.com/OCA/account-invoicing/pull/1638

Introducing a new field, discount1, and using the standard discount field to hold the result of multiple applied discounts makes more sense and reduces code complexity.

tafaRU committed:
[ADD] sale_triple_discount

chienandalu committed:
[MIG] sale_triple_discount: Migration to 10.0

OCA Transbot updated translations from Transifex

[UPD] Update sale_triple_discount.pot

thomaspot committed:
Translated using Weblate (Nederlands (nl_NL))

Currently translated at 100,0% (5 of 5 strings)

Translation: sale-workflow-10.0/sale-workflow-10.0-sale_triple_discount
Translate-URL: https://translation.odoo-community.org/projects/sale-workflow-10-0/sale-workflow-10-0-sale_triple_discount/nl_NL/

thomaspot committed:
Translated using Weblate (Dutch (nl_NL))

Currently translated at 100.0% (5 of 5 strings)

Translation: sale-workflow-10.0/sale-workflow-10.0-sale_triple_discount
Translate-URL: https://translation.odoo-community.org/projects/sale-workflow-10-0/sale-workflow-10-0-sale_triple_discount/nl_NL/

mpanarin committed:
[IMP] 10.0 sale_triple_discount: rework discount computation

jbaudoux committed:
sale_triple_discount: Fix tax total on report: include the discount2/3 in the computation

SimoRubi committed:
[FIX] sale_triple_discount: Correctly compute taxes with round_globally configuration

SimoRubi committed:
[FIX] sale_triple_discount: Store values in cache for recomputation consistency

[FIX] sale_triple_discount: Invalidating the cache of specific records before recomputing price and discount

[UPD] Update sale_triple_discount.pot

Update translation files

Updated by "Update PO files to match POT (msgmerge)" hook in Weblate.

Translation: sale-workflow-10.0/sale-workflow-10.0-sale_triple_discount
Translate-URL: https://translation.odoo-community.org/projects/sale-workflow-10-0/sale-workflow-10-0-sale_triple_discount/

grindtildeath committed:
Play onchange after invoice creation to ensure discounts are applied

grindtildeath committed:
Fix groups on sale.order view

[UPD] Update sale_triple_discount.pot

tafaRU committed:
[MIG] sale_triple_discount: Migration to 12.0

[UPD] Update sale_triple_discount.pot

Update translation files

Updated by "Update PO files to match POT (msgmerge)" hook in Weblate.

Translation: sale-workflow-12.0/sale-workflow-12.0-sale_triple_discount
Translate-URL: https://translation.odoo-community.org/projects/sale-workflow-12-0/sale-workflow-12-0-sale_triple_discount/

ps-tubtim committed:
[IMP] sale_triple_discount: black, isort, prettier

ps-tubtim committed:
[MIG] sale_triple_discount: Migration to 13.0

ferran-S73 committed:
[IMP] sale_triple_discount: black, isort, prettier

ferran-S73 committed:
[14.0][MIG] sale_triple_discount: Migration to V14

ferran-S73 committed:
[14.0][IMP] sale_triple_discount: add disc. fields in sale report templ.

ferran-S73 committed:
[14.0] sale_triple_discount: sale_order_line_discount_validation compat.

[UPD] Update sale_triple_discount.pot

[UPD] README.rst

francesco-ooops committed:
Translated using Weblate (Italian)

Currently translated at 100.0% (11 of 11 strings)

Translation: sale-workflow-14.0/sale-workflow-14.0-sale_triple_discount
Translate-URL: https://translation.odoo-community.org/projects/sale-workflow-14-0/sale-workflow-14-0-sale_triple_discount/it/

[UPD] README.rst

[UPD] README.rst

klodr committed:
Translated using Weblate (French)

Currently translated at 0.0% (0 of 11 strings)

Translation: sale-workflow-14.0/sale-workflow-14.0-sale_triple_discount
Translate-URL: https://translation.odoo-community.org/projects/sale-workflow-14-0/sale-workflow-14-0-sale_triple_discount/fr/

manuelregidor committed:
[IMP] sale_triple_discount: black, isort, prettier

manuelregidor committed:
[MIG]sale_triple_discount: Migration to 15.0

[UPD] Update sale_triple_discount.pot

[UPD] README.rst

OCA-git-bot committed:
sale_triple_discount 15.0.1.0.1

Update translation files

Updated by "Update PO files to match POT (msgmerge)" hook in Weblate.

Translation: sale-workflow-15.0/sale-workflow-15.0-sale_triple_discount
Translate-URL: https://translation.odoo-community.org/projects/sale-workflow-15-0/sale-workflow-15-0-sale_triple_discount/

sbejaoui committed:
[16.0][MIG] - sale_triple_discount

Commit:
[UPD] Update sale_triple_discount.pot

OCA-git-bot committed:
sale_triple_discount 16.0.1.0.1

OCA-git-bot committed:
[UPD] README.rst

weblate committed:
Update translation files

Updated by "Update PO files to match POT (msgmerge)" hook in Weblate.

Translation: sale-workflow-16.0/sale-workflow-16.0-sale_triple_discount
Translate-URL: https://translation.odoo-community.org/projects/sale-workflow-16-0/sale-workflow-16-0-sale_triple_discount/

alan196 committed:
Added translation using Weblate (Spanish (Mexico))

alan196 committed:
Translated using Weblate (Spanish (Mexico))

Currently translated at 100.0% (12 of 12 strings)

Translation: sale-workflow-16.0/sale-workflow-16.0-sale_triple_discount
Translate-URL: https://translation.odoo-community.org/projects/sale-workflow-16-0/sale-workflow-16-0-sale_triple_discount/es_MX/

Commit:
Translated using Weblate (Italian)

Currently translated at 100.0% (12 of 12 strings)

Translation: sale-workflow-16.0/sale-workflow-16.0-sale_triple_discount
Translate-URL: https://translation.odoo-community.org/projects/sale-workflow-16-0/sale-workflow-16-0-sale_triple_discount/it/

Ivorra78 committed:
Translated using Weblate (Spanish)

Currently translated at 100.0% (12 of 12 strings)

Translation: sale-workflow-16.0/sale-workflow-16.0-sale_triple_discount
Translate-URL: https://translation.odoo-community.org/projects/sale-workflow-16-0/sale-workflow-16-0-sale_triple_discount/es/

OCA-git-bot committed:
[UPD] README.rst

mymage committed:
Translated using Weblate (Italian)

Currently translated at 100.0% (12 of 12 strings)

Translation: sale-workflow-16.0/sale-workflow-16.0-sale_triple_discount
Translate-URL: https://translation.odoo-community.org/projects/sale-workflow-16-0/sale-workflow-16-0-sale_triple_discount/it/

renda-dev committed:
[IMP] sale_triple_discount: improve module extensibility

jappi00 committed:
Translated using Weblate (German)

Currently translated at 100.0% (12 of 12 strings)

Translation: sale-workflow-16.0/sale-workflow-16.0-sale_triple_discount
Translate-URL: https://translation.odoo-community.org/projects/sale-workflow-16-0/sale-workflow-16-0-sale_triple_discount/de/

OCA-git-bot committed:
[BOT] post-merge updates

Commit:
Translated using Weblate (Portuguese (Brazil))

Currently translated at 100.0% (12 of 12 strings)

Translation: sale-workflow-16.0/sale-workflow-16.0-sale_triple_discount
Translate-URL: https://translation.odoo-community.org/projects/sale-workflow-16-0/sale-workflow-16-0-sale_triple_discount/pt_BR/

lmignon committed:
[FIX] sale_triple_discount: Remove call to deprecated

Method 'invalidate_cache' is deprecated. To invalidate specific field's value into the cache for a recordset we must use the 'invalidate_recordset' method on the recordset itself.

OCA-git-bot committed:
[BOT] post-merge updates

lmignon committed:
[IMP] sale_triple_discount: Improves perfs and consistency

When computing the amount total on a sale.order.line, we now avoid to play with the cache since this change introduced in https://github.com/OCA/sale-workflow/commit/e3e59ba15ebcf1f91124eda60a694c0a76fabb69 introduces a decline in performance. When the cache in invalidated, fields are flushed to the database and depending computed fields are recomputed..... This ultimately leads to temporary inconsistencies breaking others addons relaying on this one. Moreover, the method used in this original commit 'self.invalidate_cache(fnames=self._discount_fields(), ids=self.ids)' is deprecated and the use of the recommended new one lead to others side effects. The original approach is replaced by a contextual method modifying only the discount field when we need it's needed to don't break code relaying on the fact that the discount field is the total discount to apply to a line. This contextual method ensures that the original value is properly restored on exit and that it will not trigger additional recompute due to the temporary change of the discount value

OCA-git-bot committed:
[BOT] post-merge updates

dalonsod committed:
[IMP] sale_triple_discount: pre-commit auto fixes

dalonsod committed:
[MIG] sale_triple_discount: Migration to 17.0

sbejaoui committed:
[MIG] sale_triple_discount: migration 18.0

sbejaoui committed:
[REF] sale_triple_discount: Consolidate discount in std field

inspired by https://github.com/OCA/account-invoicing/pull/1638

sbejaoui committed:
[FIX] sale_triple_discount: hide discount_type and disable additive type

see https://github.com/OCA/sale-workflow/issues/3649

legalsylvain commented:
Thanks for porting this module ! 

> This PR includes a refactoring inspired by https://github.com/OCA/account-invoicing/pull/1638

Did you took a look on this unfinished PR ? https://github.com/OCA/sale-workflow/pull/2885

rousseldenis commented:
/ocabot migration sale_triple_discount

rousseldenis commented:
@sbejaoui

sbejaoui commented:
> Thanks for porting this module !
> 
> > This PR includes a refactoring inspired by [OCA/account-invoicing#1638](https://github.com/OCA/account-invoicing/pull/1638)
> 
> Did you took a look on this unfinished PR ? #2885

No, but I'll have a look to see if there's anything I should take from it

lef-adhoc commented:
@sbejaoui @legalsylvain 

When there's a pricelist with a discount (visible in the UI) for the selected product on the line, it’s not being reflected in discount1. Also, regarding this, what happens if some discounts are already set and the product is changed? Ideally, it should check if any discount applies from the pricelist and update discount1 accordingly.

Can you check this module? With this PR, they both end up following the same approach.
https://github.com/ingadhoc/sale/blob/18.0/sale_three_discounts/models/sale_order_line.py
In the discount wizard we also added 3 discount fields

sbejaoui commented:
> @sbejaoui @legalsylvain
> 
> When there's a pricelist with a discount (visible in the UI) for the selected product on the line, it’s not being reflected in discount1. Also, regarding this, what happens if some discounts are already set and the product is changed? Ideally, it should check if any discount applies from the pricelist and update discount1 accordingly.
> 
> Can you check this module? With this PR, they both end up following the same approach. https://github.com/ingadhoc/sale/blob/18.0/sale_three_discounts/models/sale_order_line.py In the discount wizard we also added 3 discount fields

I had the idea to add an inverse to the standard discount field, which resets the three discounts and sets discount1 with the value coming from the inverse. this way, any standard process like pricelists will integrate with this logic out of the box, without extra customization

I will test and get back to you

legalsylvain commented:
> I had the idea to add an inverse to the standard discount field, which resets the three discounts and sets discount1 with the value coming from the inverse. this way, any standard process like pricelists will integrate with this logic out of the box, without extra customization


Very good idéa ! It should be in the mixin right ?

Let me know if there is something to review.

sbejaoui committed:
[IMP] sale_triple_discount: add inverse for standard discount field

add an inverse to the standard discount field, which resets the three
discounts and sets discount1 with the value coming from the inverse.
this way, any standard process like pricelists will integrate with this
logic out of the box, without extra customization

sbejaoui commented:
> > I had the idea to add an inverse to the standard discount field, which resets the three discounts and sets discount1 with the value coming from the inverse. this way, any standard process like pricelists will integrate with this logic out of the box, without extra customization
> 
> Very good idéa ! It should be in the mixin right ?
> 
> Let me know if there is something to review.

you can review my last commit

lef-adhoc commented:
@sbejaoui @legalsylvain @rousseldenis 
The issue is that when any of the `@depends` fields are changed, the inverse method doesn't run — so the discount isn't recalculated or updated in discount1.
If the product, product_uom, or product_uom_qty is modified, the discount should be recalculated and overwrite the previous value if applicable.
![image](https://github.com/user-attachments/assets/661ad331-2d7b-4108-99f7-f0e6293763e0)

sbejaoui commented:
> @sbejaoui @legalsylvain @rousseldenis The issue is that when any of the `@depends` fields are changed, the inverse method doesn't run — so the discount isn't recalculated or updated in discount1. If the product, product_uom, or product_uom_qty is modified, the discount should be recalculated and overwrite the previous value if applicable. 

I get what you mean. I added the compute method and ensured that the discount is recomputed when the pricelist discount applies and any of its dependencies change, check the unit test:

https://github.com/OCA/sale-workflow/blob/dfb33eb6d9231daf3db139a3847dc97bbb49e03e/sale_triple_discount/tests/test_sale_triple_discount.py#L263-L307

sbejaoui commented:
@legalsylvain , @rousseldenis 

how to fix:

`odoo.exceptions.UserError: Modules "Account Invoice Triple Discount" and "Account Fixed Discount" are incompatible.`

matteotognini commented:
Hi @sbejaoui,  could you please replace the unreleased dependencies with https://github.com/OCA/account-invoicing/pull/2019. I think that should fix the error you mentioned, as you can see in the PR conversation. Thanks!

rousseldenis commented:
> @legalsylvain , @rousseldenis
> 
> how to fix:
> 
> `odoo.exceptions.UserError: Modules "Account Invoice Triple Discount" and "Account Fixed Discount" are incompatible.

Could you add a commit to put `sale_triple_discount` and `sale_fixed_discount` in rebel modules  (with copier)?

sbejaoui committed:
[IMP] CI: add rebel modules: sale_triple_discount, sale_fixed_discount

sbejaoui commented:
> > @legalsylvain , @rousseldenis
> > how to fix:
> > `odoo.exceptions.UserError: Modules "Account Invoice Triple Discount" and "Account Fixed Discount" are incompatible.
> 
> Could you add a commit to put `sale_triple_discount` and `sale_fixed_discount` in rebel modules (with copier)?

thanks, 

now it fails at sale_order_product_availability_inline

rousseldenis commented:
> > > @legalsylvain , @rousseldenis
> > > how to fix:
> > > `odoo.exceptions.UserError: Modules "Account Invoice Triple Discount" and "Account Fixed Discount" are incompatible.
> > 
> > 
> > Could you add a commit to put `sale_triple_discount` and `sale_fixed_discount` in rebel modules (with copier)?
> 
> thanks,
> 
> now it fails at sale_order_product_availability_inline

@sbejaoui Please see : https://github.com/OCA/sale-workflow/pull/3765

lef-adhoc commented:
@sbejaoui There is an issue when changing the product quantity and a price rule is applied. The expected behavior is to set discounts 2 and 3 to 0 (which works correctly), but discount 1 should be set to the value defined by the price rule (this is how Odoo works natively—when a rule is applied, it overrides discount with the value from the pricelist).

Video:
https://drive.google.com/file/d/1tiaL8jZSlo7OboAQf0ecx3swjXG7OOrK/view

legalsylvain commented:
@lef-adhoc thanks a lot for your detailed comment.

sergioM-S73 commented:
@sbejaoui pls resolve confilcts

sbejaoui commented:
> @sbejaoui There is an issue when changing the product quantity and a price rule is applied. The expected behavior is to set discounts 2 and 3 to 0 (which works correctly), but discount 1 should be set to the value defined by the price rule (this is how Odoo works natively—when a rule is applied, it overrides discount with the value from the pricelist).
> 
> Video: https://drive.google.com/file/d/1tiaL8jZSlo7OboAQf0ecx3swjXG7OOrK/view

Thanks for the test and good catch — that was a tricky bug.
I’ve pushed a fix, could you please test it again and let me know if it works as expected?


https://github.com/user-attachments/assets/9c2f416d-0a20-4bf3-a3cc-92779a7e786e

lef-adhoc commented:
@sbejaoui 
Hi! I've been testing the feature and it works great, nice job! 👍

I wanted to propose an enhancement: what do you think about integrating the three-discount calculation into the discount wizard? I believe it would improve the user experience.

We have actually developed a working implementation of this concept in our own repository. Here is the code for the wizard, which could serve as a reference:
https://github.com/ingadhoc/sale/blob/18.0/sale_three_discounts/wizards/sale_order_discount.py

Let me know what you think!
![image](https://github.com/user-attachments/assets/cc94704c-9183-45f0-b2af-f993edf28b6c)

rousseldenis commented:
> @sbejaoui Hi! I've been testing the feature and it works great, nice job! 👍
> 
> I wanted to propose an enhancement: what do you think about integrating the three-discount calculation into the discount wizard? I believe it would improve the user experience.
> 
> We have actually developed a working implementation of this concept in our own repository. Here is the code for the wizard, which could serve as a reference: https://github.com/ingadhoc/sale/blob/18.0/sale_three_discounts/wizards/sale_order_discount.py
> 
> Let me know what you think! ![image](https://private-user-images.githubusercontent.com/148369234/463788333-cc94704c-9183-45f0-b2af-f993edf28b6c.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3NTIwNTAxOTMsIm5iZiI6MTc1MjA0OTg5MywicGF0aCI6Ii8xNDgzNjkyMzQvNDYzNzg4MzMzLWNjOTQ3MDRjLTkxODMtNDVmMC1iMmFmLWY5OTNlZGYyOGI2Yy5wbmc_WC1BbXotQWxnb3JpdGhtPUFXUzQtSE1BQy1TSEEyNTYmWC1BbXotQ3JlZGVudGlhbD1BS0lBVkNPRFlMU0E1M1BRSzRaQSUyRjIwMjUwNzA5JTJGdXMtZWFzdC0xJTJGczMlMkZhd3M0X3JlcXVlc3QmWC1BbXotRGF0ZT0yMDI1MDcwOVQwODMxMzNaJlgtQW16LUV4cGlyZXM9MzAwJlgtQW16LVNpZ25hdHVyZT1iNjJkZmYxY2QxYzViYmNlMTlhNGJhNmIyYTI1ZTA4NTRjZTg1ZjNjYjI1NTk1NTVhNjk5ZDU0ZGUwN2YxOWJjJlgtQW16LVNpZ25lZEhlYWRlcnM9aG9zdCJ9.60bbjhwgyFoAwMcseenfTILXEPwnNt7F7mB45qloQ8M)

@lef-adhoc Interesting, could you maybe make a PR on this branch in acsone repository with the improvement?

sbejaoui commented:
> @sbejaoui Hi! I've been testing the feature and it works great, nice job! 👍
> 
> I wanted to propose an enhancement: what do you think about integrating the three-discount calculation into the discount wizard? I believe it would improve the user experience.
> 
> We have actually developed a working implementation of this concept in our own repository. Here is the code for the wizard, which could serve as a reference: https://github.com/ingadhoc/sale/blob/18.0/sale_three_discounts/wizards/sale_order_discount.py
> 
> Let me know what you think! ![image](https://private-user-images.githubusercontent.com/148369234/463788333-

Thanks for the feedback, glad it’s working well!
The enhancement sounds great, but since it’s not an immediate customer requirement for us, it’s not planned for the near term
I suggest we merge this as is and open a separate PR for the improvement you proposed

ferran-S73 commented on sale_triple_discount/models/sale_order_line.py:
```diff
@@ -0,0 +1,158 @@
+
+    @api.depends("discount")
+    def _compute_discounts(self):
+        super()._compute_discount()
```
You're calling a super for a different method, is this intended?

sbejaoui replied:
yes, I will add a comment to explain why, when the discount is changed directly or via its compute method, it triggers the compute methods for the three discount fields. At this point, we need to make sure the cache is coherent before calling the inverse. calling _compute_discount() directly would alter the discount value by recomputing it from the three discounts, which may still be outdated, calling super() on the other hand avoids that

ferran-S73 commented:
@sbejaoui I've encountered a case in which the computed discount is lost.

Steps to reproduce:
 - Configure a pricelist and a pricelist item with a disocunt.
 - Set the pricelist in a partner.
 - Import via xlsx a sale order for said partner and pricelist, with a product that is included in our pricelist item.

The new sale order line created is created without any discount. I've done some code debugging and this is due to `_compute_discounts` not being triggered when sale module sets the discount from the pricelist. Thus reseting every discount to in `_get_final_discount`.

This also happens when creating a sale line from the web store, or anytime a line is created directly from calling the create method without setting any specific discount (you can try it from a server action for example).

sbejaoui commented:
> @sbejaoui I've encountered a case in which the computed discount is lost.
> 
> Steps to reproduce:
> 
> * Configure a pricelist and a pricelist item with a disocunt.
> * Set the pricelist in a partner.
> * Import via xlsx a sale order for said partner and pricelist, with a product that is included in our pricelist item.
> 
> The new sale order line created is created without any discount. I've done some code debugging and this is due to `_compute_discounts` not being triggered when sale module sets the discount from the pricelist. Thus reseting every discount to in `_get_final_discount`.
> 
> This also happens when creating a sale line from the web store, or anytime a line is created directly from calling the create method without setting any specific discount (you can try it from a server action for example).

Thank you for the test, could you please share the excel you used for the import

ferran-S73 commented:
> > @sbejaoui I've encountered a case in which the computed discount is lost.
> > Steps to reproduce:
> > 
> > * Configure a pricelist and a pricelist item with a disocunt.
> > * Set the pricelist in a partner.
> > * Import via xlsx a sale order for said partner and pricelist, with a product that is included in our pricelist item.
> > 
> > The new sale order line created is created without any discount. I've done some code debugging and this is due to `_compute_discounts` not being triggered when sale module sets the discount from the pricelist. Thus reseting every discount to in `_get_final_discount`.
> > This also happens when creating a sale line from the web store, or anytime a line is created directly from calling the create method without setting any specific discount (you can try it from a server action for example).
> 
> Thank you for the test, could you please share the excel you used for the import

Sure! Please keep in mind that I'm using some external ids and the current configuration in the runboat but that may be lost if you do another push as it is regenarated

[Sales Order (sale.order)(2).xlsx](https://github.com/user-attachments/files/21141085/Sales.Order.sale.order.2.xlsx)

ferran-S73 commented:
> > > @sbejaoui I've encountered a case in which the computed discount is lost.
> > > Steps to reproduce:
> > > 
> > > * Configure a pricelist and a pricelist item with a disocunt.
> > > * Set the pricelist in a partner.
> > > * Import via xlsx a sale order for said partner and pricelist, with a product that is included in our pricelist item.
> > > 
> > > The new sale order line created is created without any discount. I've done some code debugging and this is due to `_compute_discounts` not being triggered when sale module sets the discount from the pricelist. Thus reseting every discount to in `_get_final_discount`.
> > > This also happens when creating a sale line from the web store, or anytime a line is created directly from calling the create method without setting any specific discount (you can try it from a server action for example).
> > 
> > 
> > Thank you for the test, could you please share the excel you used for the import
> 
> Sure! Please keep in mind that I'm using some external ids and the current configuration in the runboat but that may be lost if you do another push as it is regenarated
> 
> [Sales Order (sale.order)(2).xlsx](https://github.com/user-attachments/files/21141085/Sales.Order.sale.order.2.xlsx)

Hi @sbejaoui , any news on this issue? We're currently using this pr in some clients and would like to know if you plan on giving a solution soon. It works perfectly aside from that btw.
Let me know if there's something I can do to help you solve this.

lef-adhoc commented:
Hi @sbejaoui , any progress on this ? If there is anything I can help just let me know.

akash-aipl commented on sale_triple_discount/models/sale_order_line.py:
```diff
@@ -0,0 +1,163 @@
+        self._inverse_discount()
+        return True
+
+    _sql_constraints = [
+        (
+            "discount1_limit",
+            "CHECK (discount1 <= 100.0)",
+            "Discount 1 must be lower or equal than 100%.",
+        ),
+        (
+            "discount2_limit",
+            "CHECK (discount2 <= 100.0)",
+            "Discount 2 must be lower or equal than 100%.",
+        ),
+        (
+            "discount3_limit",
+            "CHECK (discount3 <= 100.0)",
+            "Discount 3 must be lower or equal than 100%.",
+        ),
+    ]
```
I was wondering if we could consider replacing the current _sql_constraints with validation logic in create and write methods.

The reason is that using SQL constraints can cause issues during module installation or migration if there are already records with discount values above 100%. PostgreSQL would block the constraint from being added in such cases.

An alternative could be to move the validation to a helper method and call it in create and write, something like:

```

    def _check_discount_limits(self, vals):
        for field in self._discount_fields():
            if field in vals and vals[field] > 100.0:
                raise ValidationError(
                    _(
                        f"{field.replace('discount', 'Discount ').title()} must be lower than 100%."
                    )
                )

    @api.model_create_multi
    def create(self, vals_list):
        for vals in vals_list:
            self._check_discount_limits(vals)
        return super().create(vals_list)

    def write(self, vals):
        self._check_discount_limits(vals)
        return super().write(vals)

```

This would avoid migration issues while still ensuring the validation for new or updated records.
Would this approach work ?

lef-adhoc commented:
@sbejaoui Hi, , are you going to be able to continue with the PR?

lef-adhoc commented:
@sbejaoui Hi, any news on this?

pablo-cort-s73 commented:
Superseded by #3889

rousseldenis commented:
@sbejaoui
